### PR TITLE
Add Rack::Lint to RequestId tests

### PR DESCRIPTION
**Pull request by skipkayhil**
### Motivation / Background

Ref #48874

This adds additional test coverage to RequestId to validate that its input and output follow the Rack SPEC.

### Detail

In this case, the only changes necessary were to the Request tests. This is due to the fact that the Request and Response tests use different classes for their Response headers. The Response tests simulate a Rails app, where the Response headers will be a Rack::Headers object for compatbility with both Rack 2 and 3. However, since the Request tests are only using the Hash returned by the test app, the tests must use a downcased header to support both Rack 2 and Rack 3.

### Additional information

~In the future when Rack::Headers is eventually removed, `config.action_dispatch.request_id_header` will need to be changed from "X-Request-Id" to "x-request-id".~ I changed my mind and now think its important to make this change now. Is this something we should be using `new_framework_defaults` for? (Or maybe we just downcase whatever string is passed into the middleware?)

### Checklist

Before submitting the PR make sure the following are checked:

* [x] This Pull Request is related to one change. Changes that are unrelated should be opened in separate PRs.
* [x] Commit message has a detailed description of what changed and why. If this PR fixes a related issue include it in the commit message. Ex: `[Fix #issue-number]`
* [x] Tests are added or updated if you fix a bug or add a feature.
* [x] CHANGELOG files are updated for the changed libraries if there is a behavior change or additional feature. Minor bug fixes and documentation changes should not be included.

### CODE REVIEW

File: actionpack/test/dispatch/request_id_test.rb [hunk 1]
```diff
@@ -75,7 +84,9 @@ def with_test_route_set(header: "X-Request-Id")
         end
 
         @app = self.class.build_app(set) do |middleware|
+          middleware.use Rack::Lint
           middleware.use ActionDispatch::RequestId, header: header
+          middleware.use Rack::Lint
```
Comment (akhilgkrishnan) [inline actionpack/test/dispatch/request_id_test.rb:89]: Just curious to know that, Why do we need to use Rack::Lint middleware before and after the ActionDispatch::RequestId?

Comment (nunosilva800) [reply actionpack/test/dispatch/request_id_test.rb:89]: It's advised by Rack::Lint: https://github.com/rack/rack/blob/444dc8a130663159ed490f25948b73ff9b5fbba0/lib/rack/lint.rb#L27-L28

I believe this helps to ensure that the rack app, request and response are valid before it reaches your middleware, and is still valid once your middleware returns.

Comment (skipkayhil) [reply actionpack/test/dispatch/request_id_test.rb:89]: Exactly, the goal is to catch errors both going down the stack as well as up. The `Rack::Lint` after ensures that `RequestId` doesn't make any changes to the Request `env` that don't follow SPEC, and the `Rack::Lint` before ensures that `RequestId` doesn't make any changes to the Response that don't follow the SPEC.

**Comment by guilleiguaran** [comment]
Let's rebase this to include https://github.com/rails/rails/commit/ff232b1190b82a3a219f7fe92f99d2affbfb3cfd

**Pull request merged**